File: mobile/src/screens/profile/ProfileScreen.js
import { useEffect, useState } from 'react';
import { View, Text, Image, Pressable, FlatList, ActivityIndicator, StyleSheet, Linking } from 'react-native';
import { useFocusEffect } from '@react-navigation/native';
import { useCallback } from 'react';
import Screen from '../../components/common/Screen';
import Avatar from '../../components/common/Avatar';
import Button from '../../components/common/Button';
import FriendButton from '../../components/social/FriendButton';
import FollowButton from '../../components/social/FollowButton';
import PostCard from '../../components/social/PostCard';
import ActivityLog from '../../components/social/ActivityLog';
import { useAuthStore } from '../../store/authStore';
import { useAuth } from '../../hooks/useAuth';
import { fetchProfile } from '../../api/social.api';
import { fetchUserPosts } from '../../api/posts.api';
import { colors, radius } from '../../config/theme';

export default function ProfileScreen({ route, navigation }) {
  const myId = useAuthStore((s) => s.user?.id);
  const { logout } = useAuth();
  const targetId = route.params?.userId || myId;

  const [profile, setProfile] = useState(null);
  const [tab, setTab] = useState('posts');
  const [posts, setPosts] = useState(null);

  const load = useCallback(() => fetchProfile(targetId).then(setProfile), [targetId]);

  useFocusEffect(
    useCallback(() => {
      setProfile(null);
      setPosts(null);
      setTab('posts');
      load();
    }, [load])
  );

  useEffect(() => {
    if ((tab === 'posts' || tab === 'media') && profile) {
      fetchUserPosts(targetId).then(setPosts);
    }
  }, [tab, profile, targetId]);

  if (!profile) {
    return (
      <Screen>
        <View style={styles.center}>
          <ActivityIndicator color={colors.accent} />
        </View>
      </Screen>
    );
  }

  const mediaPosts = (posts || []).filter((p) => p.imageUrl);
  const totalCircle = profile.friendCount + profile.followerCount;
  const closeCirclePct = totalCircle > 0 ? Math.round((profile.friendCount / totalCircle) * 100) : 0;

  return (
    <Screen>
      <FlatList
        data={tab === 'posts' ? posts || [] : []}
        keyExtractor={(p) => p.id}
        contentContainerStyle={{ paddingBottom: 40 }}
        renderItem={({ item }) => (
          <View style={{ paddingHorizontal: 16 }}>
            <PostCard post={item} onDeleted={(id) => setPosts((prev) => prev.filter((p) => p.id !== id))} />
          </View>
        )}
        ListHeaderComponent={
          <View>
            <View style={styles.cover}>
              {profile.coverPhotoUrl ? <Image source={{ uri: profile.coverPhotoUrl }} style={styles.coverImage} /> : null}
            </View>

            <View style={styles.headerBlock}>
              <View style={styles.avatarRow}>
                <Avatar user={profile} size="lg" />
                {profile.isSelf ? (
                  <Button title="Edit profile" variant="outline" onPress={() => navigation.navigate('EditProfile')} />
                ) : (
                  <View style={{ gap: 8 }}>
                    <FriendButton profile={profile} onChange={load} />
                    <FollowButton profile={profile} onChange={load} />
                  </View>
                )}
              </View>

              <Text style={styles.name}>{profile.name}</Text>
              <Text style={styles.username}>@{profile.username}</Text>
              {profile.bio ? <Text style={styles.bio}>{profile.bio}</Text> : null}

              <View style={styles.statsRow}>
                <Text style={styles.statText}>
                  <Text style={styles.statNumber}>{profile.friendCount}</Text> friends
                </Text>
                <Text style={styles.statText}>
                  <Text style={styles.statNumber}>{profile.followerCount}</Text> followers
                </Text>
                <Text style={styles.statText}>
                  <Text style={styles.statNumber}>{profile.followingCount}</Text> following
                </Text>
              </View>

              <View style={styles.aboutCard}>
                <Text style={styles.aboutLabel}>Intro &amp; Archive Details</Text>
                {profile.about ? (
                  <View style={{ gap: 4 }}>
                    {profile.about.work ? <Text style={styles.aboutLine}>💼 {profile.about.work}</Text> : null}
                    {profile.about.education ? <Text style={styles.aboutLine}>🎓 {profile.about.education}</Text> : null}
                    {profile.about.location ? <Text style={styles.aboutLine}>📍 {profile.about.location}</Text> : null}
                    {(profile.about.links || []).map((l) => (
                      <Pressable key={l.url} onPress={() => Linking.openURL(l.url)}>
                        <Text style={[styles.aboutLine, styles.link]}>🔗 {l.label}</Text>
                      </Pressable>
                    ))}
                    {!profile.about.work && !profile.about.education && !profile.about.location && (profile.about.links || []).length === 0 && (
                      <Text style={styles.mutedText}>Nothing added yet.</Text>
                    )}
                  </View>
                ) : (
                  <Text style={styles.mutedText}>This person&apos;s info is private.</Text>
                )}
              </View>

              {totalCircle > 0 && (
                <View style={styles.aboutCard}>
                  <Text style={styles.aboutLabel}>Social Graph</Text>
                  <View style={styles.graphTrack}>
                    <View style={[styles.graphFill, { width: `${closeCirclePct}%` }]} />
                  </View>
                  <View style={styles.graphLegend}>
                    <Text style={styles.mutedText}>Friends ({closeCirclePct}%)</Text>
                    <Text style={styles.mutedText}>Followers ({100 - closeCirclePct}%)</Text>
                  </View>
                </View>
              )}

              <View style={styles.tabRow}>
                <Pressable onPress={() => setTab('posts')} style={styles.tabItem}>
                  <Text style={[styles.tabText, tab === 'posts' && styles.tabTextActive]}>Timeline &amp; Letters</Text>
                </Pressable>
                <Pressable onPress={() => setTab('media')} style={styles.tabItem}>
                  <Text style={[styles.tabText, tab === 'media' && styles.tabTextActive]}>Media &amp; Folio</Text>
                </Pressable>
                {profile.isSelf && (
                  <Pressable onPress={() => setTab('activity')} style={styles.tabItem}>
                    <Text style={[styles.tabText, tab === 'activity' && styles.tabTextActive]}>Activity Log</Text>
                  </Pressable>
                )}
              </View>

              {profile.isSelf && (
                <Pressable onPress={logout} style={{ marginTop: 16 }}>
                  <Text style={styles.logout}>Log out</Text>
                </Pressable>
              )}
            </View>

            {tab === 'posts' && posts === null && <ActivityIndicator style={{ marginTop: 20 }} color={colors.accent} />}
            {tab === 'posts' && posts?.length === 0 && <Text style={styles.mutedCentered}>No posts yet.</Text>}

            {tab === 'media' && (
              <View style={styles.mediaGrid}>
                {mediaPosts.length === 0 ? (
                  <Text style={styles.mutedCentered}>No photos yet.</Text>
                ) : (
                  mediaPosts.map((p) => <Image key={p.id} source={{ uri: p.imageUrl }} style={styles.mediaImage} />)
                )}
              </View>
            )}

            {tab === 'activity' && (
              <View style={{ paddingHorizontal: 16 }}>
                <ActivityLog />
              </View>
            )}
          </View>
        }
      />
    </Screen>
  );
}

const styles = StyleSheet.create({
  center: { flex: 1, alignItems: 'center', justifyContent: 'center' },
  cover: { height: 140, backgroundColor: colors.paperSoft },
  coverImage: { width: '100%', height: '100%' },
  headerBlock: { paddingHorizontal: 16, marginTop: -32 },
  avatarRow: { flexDirection: 'row', justifyContent: 'space-between', alignItems: 'flex-end', marginBottom: 10 },
  name: { fontSize: 20, fontWeight: '800', color: colors.ink },
  username: { fontSize: 13, color: colors.inkSoft },
  bio: { fontSize: 13, color: colors.ink, marginTop: 6 },
  statsRow: { flexDirection: 'row', gap: 16, marginTop: 10 },
  statText: { fontSize: 13, color: colors.inkSoft },
  statNumber: { fontWeight: '800', color: colors.ink },
  aboutCard: { backgroundColor: colors.paper, borderRadius: radius.xl, padding: 14, marginTop: 12 },
  aboutLabel: { fontSize: 11, fontWeight: '700', color: colors.inkSoft, textTransform: 'uppercase', marginBottom: 8 },
  aboutLine: { fontSize: 13, color: colors.ink },
  link: { color: colors.accent, textDecorationLine: 'underline' },
  mutedText: { fontSize: 13, color: colors.inkSoft },
  mutedCentered: { fontSize: 13, color: colors.inkSoft, textAlign: 'center', marginTop: 20 },
  graphTrack: { height: 8, borderRadius: 4, backgroundColor: colors.accentSoft, overflow: 'hidden', marginBottom: 6 },
  graphFill: { height: '100%', backgroundColor: colors.accent },
  graphLegend: { flexDirection: 'row', justifyContent: 'space-between' },
  tabRow: { flexDirection: 'row', gap: 16, borderBottomWidth: 1, borderBottomColor: colors.line, marginTop: 16, paddingBottom: 10 },
  tabItem: {},
  tabText: { fontSize: 13, color: colors.inkSoft, fontWeight: '600' },
  tabTextActive: { color: colors.accent, fontWeight: '800' },
  mediaGrid: { flexDirection: 'row', flexWrap: 'wrap', gap: 6, padding: 16 },
  mediaImage: { width: '31.5%', aspectRatio: 1, borderRadius: radius.sm },
  logout: { color: colors.inkSoft, fontSize: 13 },
});

File: mobile/lib/features/profile/screens/profile_screen.dart
import 'package:cached_network_image/cached_network_image.dart';
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:go_router/go_router.dart';
import 'package:url_launcher/url_launcher.dart';

import '../../../common/widgets/app_button.dart';
import '../../../common/widgets/avatar.dart';
import '../../../common/widgets/screen.dart';
import '../../../config/env.dart';
import '../../../config/theme.dart';
import '../../../core/auth_session.dart';
import '../../../models/post.dart';
import '../../../models/profile.dart';
import '../../auth/providers/auth_provider.dart';
import '../../chat/providers/chat_actions.dart';
import '../../feed/data/posts_api.dart';
import '../data/block_api.dart';
import '../data/social_api.dart';
import '../widgets/activity_log.dart';
import '../widgets/follow_button.dart';
import 'follow_list_screen.dart';

enum _ProfileTab { posts, media, activity }

/// Ports `src/screens/profile/ProfileScreen.js`: posts/media/(activity)
/// tabs, cover+avatar header, About card, friend/follow actions.
///
/// Deviation from the RN app: RN refetches on every screen *focus* via
/// `useFocusEffect` (React Navigation keeps every screen instance around).
/// go_router's StatefulShellRoute also keeps this screen's Navigator alive
/// across tab switches, so refetch-on-focus isn't available the same way —
/// this fetches once on mount instead and offers pull-to-refresh.
class ProfileScreen extends ConsumerStatefulWidget {
  final String? userId;

  const ProfileScreen({super.key, this.userId});

  @override
  ConsumerState<ProfileScreen> createState() => _ProfileScreenState();
}

class _ProfileScreenState extends ConsumerState<ProfileScreen> {
  UserProfile? _profile;
  List<Post>? _posts;
  _ProfileTab _tab = _ProfileTab.posts;
  bool _loadingPosts = false;

  String get _targetId =>
      widget.userId ?? ref.read(authSessionProvider).userId!;

  @override
  void initState() {
    super.initState();
    _load();
  }

  Future<void> _load() async {
    final profile = await ref.read(socialApiProvider).fetchProfile(_targetId);
    if (!mounted) return;
    setState(() => _profile = profile);
    await _loadPosts();
  }

  Future<void> _loadPosts() async {
    setState(() => _loadingPosts = true);
    try {
      final posts = await ref.read(postsApiProvider).fetchUserPosts(_targetId);
      if (mounted) setState(() => _posts = posts);
    } finally {
      if (mounted) setState(() => _loadingPosts = false);
    }
  }

  @override
  Widget build(BuildContext context) {
    final profile = _profile;

    return Screen(
      child: Scaffold(
        backgroundColor: Colors.transparent,
        appBar: AppBar(title: const Text('Profile')),
        body: profile == null
            ? const Center(
                child: CircularProgressIndicator(color: AppColors.accent),
              )
            : RefreshIndicator(
                onRefresh: _load,
                child: ListView(
                  padding: EdgeInsets.zero,
                  children: [
                    _Header(profile: profile, onChange: _load),
                    _StatsRow(profile: profile, userId: _targetId),
                    _AboutCard(about: profile.about),
                    _TabBarRow(
                      tab: _tab,
                      showActivity: profile.isSelf,
                      onSelect: (t) => setState(() => _tab = t),
                    ),
                    const SizedBox(height: 8),
                    _TabContent(
                      tab: _tab,
                      posts: _posts,
                      loading: _loadingPosts,
                    ),
                    const SizedBox(height: 32),
                  ],
                ),
              ),
      ),
    );
  }
}

class _Header extends ConsumerStatefulWidget {
  final UserProfile profile;
  final VoidCallback onChange;

  const _Header({required this.profile, required this.onChange});

  @override
  ConsumerState<_Header> createState() => _HeaderState();
}

class _HeaderState extends ConsumerState<_Header> {
  bool _startingChat = false;

  Future<void> _startChat() async {
    setState(() => _startingChat = true);
    try {
      final conversation = await ref
          .read(chatActionsProvider)
          .startDirectConversation(widget.profile.id);
      if (mounted) context.push('/chat/${conversation.id}');
    } finally {
      if (mounted) setState(() => _startingChat = false);
    }
  }

  Future<void> _toggleBlock() async {
    final api = ref.read(blockApiProvider);
    if (widget.profile.isBlocked) {
      await api.unblockUser(widget.profile.id);
    } else {
      await api.blockUser(widget.profile.id);
    }
    widget.onChange();
  }

  @override
  Widget build(BuildContext context) {
    final profile = widget.profile;
    final onChange = widget.onChange;
    final myId = ref.watch(authSessionProvider).userId;
    final isSelf = profile.isSelf || profile.id == myId;

    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        Container(
          height: 140,
          width: double.infinity,
          color: AppColors.paperSoft,
          child: profile.coverPhotoUrl != null
              ? CachedNetworkImage(
                  imageUrl: Env.resolveMediaUrl(profile.coverPhotoUrl),
                  fit: BoxFit.cover,
                )
              : null,
        ),
        Padding(
          padding: const EdgeInsets.symmetric(horizontal: 16),
          child: Transform.translate(
            offset: const Offset(0, -32),
            child: Row(
              crossAxisAlignment: CrossAxisAlignment.end,
              children: [
                Avatar(user: profile.asUser(), size: AvatarSize.lg),
                const SizedBox(width: 12),
                Expanded(
                  child: Padding(
                    padding: const EdgeInsets.only(bottom: 8),
                    child: Column(
                      crossAxisAlignment: CrossAxisAlignment.start,
                      children: [
                        Text(
                          profile.name,
                          style: const TextStyle(
                            fontSize: 18,
                            fontWeight: FontWeight.w700,
                          ),
                        ),
                        Text(
                          '@${profile.username}',
                          style: const TextStyle(
                            color: AppColors.inkSoft,
                            fontSize: 13,
                          ),
                        ),
                      ],
                    ),
                  ),
                ),
                Padding(
                  padding: const EdgeInsets.only(bottom: 8),
                  child: isSelf
                      ? AppButton(
                          label: 'Edit profile',
                          variant: AppButtonVariant.outline,
                          onPressed: () => context.push('/profile/edit'),
                        )
                      : profile.blockedByOther
                      ? const SizedBox.shrink()
                      : Row(
                          children: [
                            FollowButton(profile: profile, onChange: onChange),
                            if (profile.isMutual) ...[
                              const SizedBox(width: 8),
                              AppButton(
                                label: 'Chat',
                                loading: _startingChat,
                                onPressed: _startingChat ? null : _startChat,
                              ),
                            ],
                            const SizedBox(width: 4),
                            PopupMenuButton<String>(
                              icon: const Icon(
                                Icons.more_vert,
                                color: AppColors.inkSoft,
                              ),
                              onSelected: (value) {
                                if (value == 'block') _toggleBlock();
                              },
                              itemBuilder: (context) => [
                                PopupMenuItem(
                                  value: 'block',
                                  child: Text(
                                    profile.isBlocked ? 'Unblock' : 'Block',
                                    style: const TextStyle(
                                      color: AppColors.danger,
                                    ),
                                  ),
                                ),
                              ],
                            ),
                          ],
                        ),
                ),
              ],
            ),
          ),
        ),
        if (profile.bio != null && profile.bio!.isNotEmpty)
          Padding(
            padding: const EdgeInsets.symmetric(horizontal: 16),
            child: Text(profile.bio!, style: const TextStyle(fontSize: 14)),
          ),
        if (isSelf)
          Padding(
            padding: const EdgeInsets.fromLTRB(16, 12, 16, 0),
            child: Align(
              alignment: Alignment.centerLeft,
              child: TextButton(
                onPressed: () => ref.read(authActionsProvider).logout(),
                child: const Text(
                  'Log out',
                  style: TextStyle(color: AppColors.danger),
                ),
              ),
            ),
          ),
      ],
    );
  }
}

class _StatsRow extends StatelessWidget {
  final UserProfile profile;
  final String userId;

  const _StatsRow({required this.profile, required this.userId});

  @override
  Widget build(BuildContext context) {
    // Followers/following are only browsable for your own profile or once
    // you're mutually followed — same gate as the web app's ProfilePage.
    final canSeeLists = profile.isSelf || profile.isMutual;

    Widget stat(String label, int value, FollowListType type) {
      final child = Column(
        children: [
          Text(
            '$value',
            style: const TextStyle(fontSize: 16, fontWeight: FontWeight.w700),
          ),
          Text(
            label,
            style: const TextStyle(fontSize: 12, color: AppColors.inkSoft),
          ),
        ],
      );
      if (!canSeeLists) return child;
      return InkWell(
        borderRadius: BorderRadius.circular(AppRadius.sm),
        onTap: () => Navigator.of(context).push(
          MaterialPageRoute(
            builder: (_) => FollowListScreen(userId: userId, type: type),
          ),
        ),
        child: Padding(padding: const EdgeInsets.all(4), child: child),
      );
    }

    return Padding(
      padding: const EdgeInsets.symmetric(vertical: 8),
      child: Row(
        mainAxisAlignment: MainAxisAlignment.spaceEvenly,
        children: [
          stat('Followers', profile.followerCount, FollowListType.followers),
          stat('Following', profile.followingCount, FollowListType.following),
        ],
      ),
    );
  }
}

class _AboutCard extends StatelessWidget {
  final AboutInfo? about;

  const _AboutCard({required this.about});

  @override
  Widget build(BuildContext context) {
    Widget card(Widget child) => Container(
      margin: const EdgeInsets.symmetric(horizontal: 16, vertical: 8),
      padding: const EdgeInsets.all(14),
      decoration: BoxDecoration(
        color: AppColors.paper,
        borderRadius: BorderRadius.circular(AppRadius.md),
        border: Border.all(color: AppColors.line),
      ),
      child: child,
    );

    if (about == null) {
      return card(
        const Text(
          "This person's info is private.",
          style: TextStyle(color: AppColors.inkSoft),
        ),
      );
    }
    if (about!.isEmpty) {
      return card(
        const Text(
          'Nothing added yet.',
          style: TextStyle(color: AppColors.inkSoft),
        ),
      );
    }

    return card(
      Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          if (about!.work != null) Text('💼 ${about!.work}'),
          if (about!.education != null) Text('🎓 ${about!.education}'),
          if (about!.location != null) Text('📍 ${about!.location}'),
          for (final link in about!.links)
            InkWell(
              onTap: () => launchUrl(
                Uri.parse(link.url),
                mode: LaunchMode.externalApplication,
              ),
              child: Padding(
                padding: const EdgeInsets.only(top: 4),
                child: Text(
                  '🔗 ${link.label}',
                  style: const TextStyle(color: AppColors.accent),
                ),
              ),
            ),
        ],
      ),
    );
  }
}

class _TabBarRow extends StatelessWidget {
  final _ProfileTab tab;
  final bool showActivity;
  final ValueChanged<_ProfileTab> onSelect;

  const _TabBarRow({
    required this.tab,
    required this.showActivity,
    required this.onSelect,
  });

  @override
  Widget build(BuildContext context) {
    Widget chip(String label, _ProfileTab value) {
      final active = tab == value;
      return Padding(
        padding: const EdgeInsets.only(right: 8),
        child: ChoiceChip(
          label: Text(label),
          selected: active,
          onSelected: (_) => onSelect(value),
          selectedColor: AppColors.accentSoft,
          labelStyle: TextStyle(
            color: active ? AppColors.accent : AppColors.inkSoft,
          ),
        ),
      );
    }

    return Padding(
      padding: const EdgeInsets.symmetric(horizontal: 16),
      child: Row(
        children: [
          chip('Posts', _ProfileTab.posts),
          chip('Media', _ProfileTab.media),
          if (showActivity) chip('Activity', _ProfileTab.activity),
        ],
      ),
    );
  }
}

class _TabContent extends StatelessWidget {
  final _ProfileTab tab;
  final List<Post>? posts;
  final bool loading;

  const _TabContent({
    required this.tab,
    required this.posts,
    required this.loading,
  });

  @override
  Widget build(BuildContext context) {
    if (tab == _ProfileTab.activity) {
      return const ActivityLog();
    }

    if (loading && posts == null) {
      return const Padding(
        padding: EdgeInsets.all(24),
        child: Center(child: CircularProgressIndicator()),
      );
    }

    final items = posts ?? [];
    if (tab == _ProfileTab.media) {
      final media = items.where((p) => p.imageUrl != null).toList();
      if (media.isEmpty) {
        return const Padding(
          padding: EdgeInsets.all(24),
          child: Center(
            child: Text(
              'No media yet.',
              style: TextStyle(color: AppColors.inkSoft),
            ),
          ),
        );
      }
      return GridView.builder(
        padding: const EdgeInsets.symmetric(horizontal: 16),
        shrinkWrap: true,
        physics: const NeverScrollableScrollPhysics(),
        gridDelegate: const SliverGridDelegateWithFixedCrossAxisCount(
          crossAxisCount: 3,
          crossAxisSpacing: 4,
          mainAxisSpacing: 4,
        ),
        itemCount: media.length,
        itemBuilder: (context, i) => ClipRRect(
          borderRadius: BorderRadius.circular(AppRadius.sm),
          child: CachedNetworkImage(
            imageUrl: Env.resolveMediaUrl(media[i].imageUrl),
            fit: BoxFit.cover,
          ),
        ),
      );
    }

    if (items.isEmpty) {
      return const Padding(
        padding: EdgeInsets.all(24),
        child: Center(
          child: Text(
            'No posts yet.',
            style: TextStyle(color: AppColors.inkSoft),
          ),
        ),
      );
    }

    return Column(
      children: items
          .map(
            (post) => Container(
              margin: const EdgeInsets.fromLTRB(16, 0, 16, 12),
              padding: const EdgeInsets.all(14),
              decoration: BoxDecoration(
                color: AppColors.paper,
                borderRadius: BorderRadius.circular(AppRadius.md),
                border: Border.all(color: AppColors.line),
              ),
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  if (post.content != null) Text(post.content!),
                  if (post.imageUrl != null) ...[
                    const SizedBox(height: 8),
                    ClipRRect(
                      borderRadius: BorderRadius.circular(AppRadius.sm),
                      child: CachedNetworkImage(
                        imageUrl: Env.resolveMediaUrl(post.imageUrl),
                        fit: BoxFit.cover,
                      ),
                    ),
                  ],
                ],
              ),
            ),
          )
          .toList(),
    );
  }
}
